File: Week6 (Graph [SSSP, APSP, BFS, DFS])/Single Source Shortest Path (SSSP)/lbh_bellman-ford.py
class Graph:
    def __init__(self, vertices):
        self.V = vertices
        self.edges = []

    def add_edge(self, u, v, w):
        self.edges.append([u, v, w])

    def print_solution(self, dist):
        print("Vertex \t Distance from Source")
        for i in range(self.V):
            print(f"{i}\t\t{dist[i]}")

    def print_path(self, parent, vertex):
        if parent[vertex] == -1:
            print(vertex, end=" ")
            return
        self.print_path(parent, parent[vertex])
        print(vertex, end=" ")

    def bellman_ford(self, src):
        dist = [float("Inf")] * self.V
        parent = [-1] * self.V

        dist[src] = 0

        # Relax all edges V - 1 times
        for _ in range(self.V - 1):
            for u, v, w in self.edges:
                if dist[u] != float("Inf") and dist[u] + w < dist[v]:
                    dist[v] = dist[u] + w
                    parent[v] = u

        # Check for negative-weight cycles
        for u, v, w in self.edges:
            if dist[u] != float("Inf") and dist[u] + w < dist[v]:
                print("Graph contains negative weight cycle")
                return

        self.print_solution(dist)

        # Example: Print the path from vertex 0 to vertex 3
        print("\nShortest path from 0 to 3: ", end="")
        if dist[3] != float("Inf"):
            self.print_path(parent, 3)
        else:
            print("No path exists.")
        print()

# Example usage
g = Graph(5)
g.add_edge(0, 1, -1)
g.add_edge(0, 2, 4)
g.add_edge(1, 2, 3)
g.add_edge(1, 3, 2)
g.add_edge(1, 4, 2)
g.add_edge(3, 2, 5)
g.add_edge(3, 1, 1)
g.add_edge(4, 3, -3)

g.bellman_ford(0)

File: Week6 (Graph [SSSP, APSP, BFS, DFS])/Single Source Shortest Path (SSSP)/lbh_bellman-ford.cpp
#include <iostream>
#include <vector>
#include <climits>
using namespace std;

struct Edge {
    int src, dest, weight;
};

// Function to print the final shortest distances
void printSolution(const vector<int>& dist) {
    cout << "Vertex \t Distance from Source\n";
    for (int i = 0; i < dist.size(); i++)
        cout << i << " \t " << dist[i] << endl;
}

// Function to print the shortest path from source to vertex
void printPath(const vector<int>& parent, int vertex) {
    if (parent[vertex] == -1) {
        cout << vertex << " ";
        return;
    }
    printPath(parent, parent[vertex]);
    cout << vertex << " ";
}

// Bellman-Ford algorithm with path reconstruction
void BellmanFord(const vector<Edge>& edges, int V, int E, int src) {
    vector<int> dist(V, INT_MAX);
    vector<int> parent(V, -1);

    dist[src] = 0;

    // Relax all edges |V| - 1 times
    for (int i = 1; i <= V - 1; i++) {
        for (int j = 0; j < E; j++) {
            int u = edges[j].src;
            int v = edges[j].dest;
            int weight = edges[j].weight;
            if (dist[u] != INT_MAX && dist[u] + weight < dist[v]) {
                dist[v] = dist[u] + weight;
                parent[v] = u;
            }
        }
    }

    // Check for negative-weight cycles
    for (int i = 0; i < E; i++) {
        int u = edges[i].src;
        int v = edges[i].dest;
        int weight = edges[i].weight;
        if (dist[u] != INT_MAX && dist[u] + weight < dist[v]) {
            cout << "Graph contains negative weight cycle\n";
            return;
        }
    }

    printSolution(dist);

    // Example: Print the path from vertex 0 to vertex 3
    cout << "\nShortest path from 0 to 3: ";
    if (dist[3] != INT_MAX) {
        printPath(parent, 3);
    } else {
        cout << "No path exists.";
    }
    cout << endl;
}

int main() {
    vector<Edge> edges = {
        {0, 1, -1}, {0, 2, 4}, {1, 2, 3}, {1, 3, 2}, {1, 4, 2},
        {3, 2, 5}, {3, 1, 1}, {4, 3, -3}
    };

    BellmanFord(edges, 5, 8, 0);

    return 0;
}